esrally/track/params.py

```diff
@@ -953,39 +953,54 @@ def create_readers(
     create_reader,
 ):
     logger = logging.getLogger(__name__)
-    # pre-initialize in order to assign parallel tasks for different corpora through assignment
-    readers = [None for corpus in corpora for _ in corpus.documents] * num_clients
+    readers = []
+    total_readers = 0
     # stagger which corpus each client starts with for better parallelism
-    for group, corpus in enumerate(corpora[(start_client_index + mod) % len(corpora)] for mod in range(len(corpora))):
-        for entry, docs in enumerate(corpus.documents):
+    reordered_corpora = corpora[start_client_index:] + corpora[:start_client_index]
+    for corpus in reordered_corpora:
+        readers.append(collections.deque())
+        for docs in corpus.documents:
             offset, num_docs, num_lines = bounds(
                 docs.number_of_documents, start_client_index, end_client_index, num_clients, docs.includes_action_and_meta_data
             )
-            if num_docs > 0:
-                target = f"{docs.target_index}/{docs.target_type}" if docs.target_index else "/"
-                if docs.target_data_stream:
-                    target = docs.target_data_stream
-                logger.debug(
-                    "Task-relative clients at index [%d-%d] will bulk index [%d] docs starting from line offset [%d] for [%s] "
-                    "from corpus [%s].",
-                    start_client_index,
-                    end_client_index,
-                    num_docs,
-                    offset,
-                    target,
-                    corpus.name,
-                )
-                readers[len(corpora) * entry + group] = create_reader(
-                    docs, offset, num_lines, num_docs, batch_size, bulk_size, id_conflicts, conflict_probability, on_conflict, recency
-                )
-            else:
+            if num_docs == 0:
                 logger.debug(
                     "Task-relative clients at index [%d-%d] skip [%s] (no documents to read).",
                     start_client_index,
                     end_client_index,
                     corpus.name,
                 )
-    return readers
+                continue
+
+            target = f"{docs.target_index}/{docs.target_type}" if docs.target_index else "/"
```

Comment:
This and the following two lines add a bit of noise in a quite complicated function. Maybe we can help the reader by moving this into a separate function?

Reply:
So this was only needed for the debug logs, which gave me one more reason to remove them, which I have done in [`a26e18c` (#1659)](https://github.com/elastic/rally/pull/1659/commits/a26e18c84092601c1f116bbbc11b05a81a9585ce)